pydeseq2/dds.py

```diff
@@ -213,55 +226,153 @@ def __init__(
             raise ValueError(
                 "Either adata or both counts and metadata arguments must be provided."
             )
-
-        # Convert design_factors to list if a single string was provided.
-        self.design_factors = (
-            [design_factors] if isinstance(design_factors, str) else design_factors
-        )
-        self.continuous_factors = continuous_factors
-
-        if self.obs[self.design_factors].isna().any().any():
-            raise ValueError("NaNs are not allowed in the design factors.")
-        self.obs[self.design_factors] = self.obs[self.design_factors].astype(str)
+        for col in self.obs.columns:
+            if col.strip() != col:
+                raise ValueError("Columns cannot contain leading or trailing spaces")
 
         # Check that design factors don't contain underscores. If so, convert them to
         # hyphens.
-        if np.any(["_" in factor for factor in self.design_factors]):
+        self.design_factors = design_factors
+
+        # Following lines handle the case where the user provides a formula
+        # of the type formulaic can recognize or a column
+        if isinstance(self.design_factors, str):
+            self.continuous_factors = continuous_factors
+            if self.design_factors not in self.obs.columns:
+                # TODO uncomment when we support all of formulaic grammar
+                # try:
+                #     Formula(self.design_factors)
+                # except FormulaInvalidError as err:
+                #     raise (
+                #         f"Design factor {self.design_factors} is not a valid formula"
+                #     ) from err
+                # Remove potential spaces around + signs
+                self.design_factors = re.sub(r"\ \+|\+\ ", "+", self.design_factors)
+                # Check that we have something like ~a + b + a:b:c:d
+                assert self.design_factors.startswith("~")
+
+                for factor in itertools.chain(
+                    *[term.split(":") for term in self.design_factors[1:].split("+")]
+                ):
+                    if factor not in self.obs.columns:
+                        raise ValueError(f"Formula contain unknown factor {factor}")
+
+                self.single_design_factors = list(
+                    {col for col in self.obs.columns if col in self.design_factors}
+                    & set(self.obs.columns)
+                )
+            else:
+                # Only one factor
+                self.design_factors = [self.design_factors]
+                self.single_design_factors = self.design_factors
```

Comment:
If it simplifies the code we could update the convention regarding  the `design_factors` argument and decide that

- if it's a string, it has to be a valid `formulaic` formula (otherwise we throw an error)
- otherwise, it has to be a list of strings (each of which is a column of the metadata)

I'm wondering if accepting strings both as formulas and single design factors wouldn't be confusing. Besides, with this new convention the old `design_factors = "condition"` could still be easily converted into a formula: `design_factors = "~condition"`.

Reply:
Ok let's simplify !